File: 2025_go/11_reactor.py
#!/usr/bin/env python3

from functools import lru_cache


def solve11(lines: list[str], b: bool) -> int:
    nodes: dict[str, list[str]] = {}
    for line in lines:
        parts = line.split()
        key = parts[0][:-1]
        nodes[key] = parts[1:]

    @lru_cache(maxsize=None)
    def dfs(node: str, has_fft: bool, has_dac: bool) -> int:
        has_fft |= node == "fft"
        has_dac |= node == "dac"
        total = 0
        for next_node in nodes.get(node, []):
            if next_node == "out":
                if not b or (has_fft and has_dac):
                    total += 1
            else:
                total += dfs(next_node, has_fft, has_dac)
        return total
    start = "svr" if b else "you"
    return dfs(start, False, False)


if __name__ == "__main__":
    with open("inputs/11_sample_a.txt") as f:
        sa = [l for line in f if (l := line.strip())]
    with open("inputs/11_sample_b.txt") as f:
        sb = [l for line in f if (l := line.strip())]
    with open("inputs/11_input.txt") as f:
        i = [l for line in f if (l := line.strip())]

    print(f"Sample A: {solve11(sa, False)}")  # 5
    print(f"A:        {solve11(i, False)}")   # 599
    print(f"Sample B: {solve11(sb, True)}")   # 2
    print(f"B:        {solve11(i, True)}")    # 393474305030400

File: 2025_go/11_reactor.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func main() {
	fmt.Println("Sample A", solve11("inputs/11_sample_a.txt", false)) // 5
	fmt.Println("A", solve11("inputs/11_input.txt", false))           // 599
	fmt.Println("Sample B", solve11("inputs/11_sample_b.txt", true))  // 2
	fmt.Println("B", solve11("inputs/11_input.txt", true))            // 393474305030400
}

type state struct {
	node string
	fft  bool
	dac  bool
}

func solve11(filename string, b bool) int {
	var (
		f, _    = os.Open(filename)
		scanner = bufio.NewScanner(f)
		nodes   = make(map[string][]string)
		cache   = make(map[state]int)
	)
	defer f.Close()
	for scanner.Scan() {
		parts := strings.Fields(scanner.Text())
		key := parts[0][:len(parts[0])-1]
		for i := 1; i < len(parts); i++ {
			nodes[key] = append(nodes[key], parts[i])
		}
	}
	start := "you"
	if b {
		start = "svr"
	}
	return dfs(start, false, false, b, nodes, cache)
}

func dfs(node string, fft, dac, b bool, nodes map[string][]string, cache map[state]int) int {
	if node == "fft" {
		fft = true
	}
	if node == "dac" {
		dac = true
	}
	s := state{node, fft, dac}
	if v, ok := cache[s]; ok {
		return v
	}
	total := 0
	for _, next := range nodes[node] {
		if next == "out" {
			if !b || (fft && dac) {
				total++
			}
		} else {
			total += dfs(next, fft, dac, b, nodes, cache)
		}
	}
	cache[s] = total
	return total
}
